### PR TITLE
fix message overlap

### PR DESCRIPTION
**Pull request by chrisnojima**
So we've had a cellCache for a long time so we could do rowIndex to messagekey mapping for heights, but it turns out this is used by a cellRangeRenderer which also has caching at multiple levels (style, cellsize etc). so it turns out its better to just implement that thing. We also now set the key to be rowIndex-messageCount so if the count isn't changing we can reuse keys better and if we do shuffle things around we can make sure the caches are invalidated

### CODE REVIEW

File: shared/chat/conversation/list.desktop.js [hunk 1]
```diff
@@ -31,6 +31,7 @@ type State = {
 const scrollbarWidth = 20
 const lockedToBottomSlop = 20
 const listBottomMargin = 10
+const DEBUG_ROW_RENDER = __DEV__ && false
```
Comment (chrisnojima) [inline shared/chat/conversation/list.desktop.js]: set this to true to see coloring and overlap off, helps debug visual issues

File: shared/chat/conversation/list.desktop.js [hunk 2]
```diff
@@ -50,7 +51,10 @@ class ConversationList extends Component<void, Props, State> {
       scrollTop: 0,
     }
 
-    this._cellCache = new CellSizeCache(this._indexToID)
+    this._cellCache = new DefaultCellMeasurerCellSizeCache({
```
Comment (chrisnojima) [inline shared/chat/conversation/list.desktop.js]: we use the default version of this now

File: shared/chat/conversation/list.desktop.js [hunk 3]
```diff
@@ -117,10 +121,12 @@ class ConversationList extends Component<void, Props, State> {
 
     // If we're not scrolling let's update our internal messages
     if (!this.state.isScrolling) {
-      this._invalidateChangedMessages(nextProps)
-      this.setState({
-        messages: nextProps.messages,
-      })
+      if (nextProps.messages !== this.state.messages) {
```
Comment (chrisnojima) [inline shared/chat/conversation/list.desktop.js]: Update the  state only if it changed

File: shared/chat/conversation/list.desktop.js [hunk 4]
```diff
@@ -305,6 +319,8 @@ class ConversationList extends Component<void, Props, State> {
     this._list = r
   }
 
+  _cellRangeRenderer = options => chatCellRangeRenderer(this.state.messages.count(), this._cellCache, options)
```
Comment (chrisnojima) [inline shared/chat/conversation/list.desktop.js]: pass through some extra stuff into the function

File: shared/chat/conversation/list.desktop.js [hunk 5]
```diff
@@ -389,37 +406,128 @@ const listStyle = {
...
 }
 
-class CellSizeCache extends defaultCellMeasurerCellSizeCache {
-  _indexToID: (index: number) => any;
-
-  constructor (indexToID) {
-    super({uniformColumnWidth: true, uniformRowHeight: false})
-    this._indexToID = indexToID
+let lastMessageCount
+function chatCellRangeRenderer (messageCount: number, cellSizeCache: any, {
```
Comment (chrisnojima) [inline shared/chat/conversation/list.desktop.js]: a copy/paste fork of https://github.com/bvaughn/react-virtualized/blob/master/source/Grid/defaultCellRangeRenderer.js#L1

File: shared/chat/conversation/list.desktop.js [hunk 6]
```diff
@@ -389,37 +406,128 @@ const listStyle = {
...
+  styleCache,
+  verticalOffsetAdjustment,
+  visibleColumnIndices,
+  visibleRowIndices,
+}: DefaultCellRangeRendererParams) {
+  const renderedCells = []
+  const offsetAdjusted = verticalOffsetAdjustment || horizontalOffsetAdjustment
+  const canCacheStyle = !isScrolling || !offsetAdjusted
+
+  if (messageCount !== lastMessageCount) {
```
Comment (chrisnojima) [inline shared/chat/conversation/list.desktop.js]: toss our caches if the count changes

File: shared/desktop/client.js [hunk 7]
```diff
@@ -14,7 +14,8 @@ const name = path.join(__dirname, 'dist', 'main.bundle.js')
 const params = [name]
 
 // Find extensions
-let devToolRoots = process.env.KEYBASE_DEV_TOOL_ROOTS
+
+let devToolRoots = !process.env.KEYBASE_PERF && process.env.KEYBASE_DEV_TOOL_ROOTS
```
Comment (chrisnojima) [inline shared/desktop/client.js]: don't load this stuff when doing perf

File: shared/package.json [hunk 8]
```diff
@@ -75,7 +75,7 @@
     "react-native-push-notification": "2.2.1",
     "react-redux": "5.0.1",
     "react-tap-event-plugin": "2.0.1",
-    "react-virtualized": "8.8.1",
+    "react-virtualized": "8.11.0",
```
Comment (chrisnojima) [inline shared/package.json]: update virtualized

**Comment by chrisnojima** [comment]
@keybase/react-hackers

**Comment by MarcoPolo** [comment]
I'm reviewing this now

**Pull request merged**